addon/wait-for-promise.ts

```diff
@@ -23,24 +24,83 @@ const PROMISE_WAITER = new TestWaiter<Promise<unknown>>('promise-waiter');
  *     }));
  *   }
  * }
+ *
+ * @example
+ *
+ * import Component from '@ember/component';
+ * import { waitForPromise } from 'ember-test-waiters';
+ *
+ * export default class Friendz extends Component {
+ *   @waitForPromise
+ *   async doAsyncStuff() {
+ *     await somethingAsync();
+ *   }
+ * }
+ *
+ * @example
+ *
+ * import Component from '@ember/component';
+ * import { waitForPromise } from 'ember-test-waiters';
+ *
+ * export default Component.extend({
+ *   doAsyncStuff: waitForPromise(async function doAsyncStuff() {
+ *     await somethingAsync();
+ *   }
+ * });
  */
-export default function waitForPromise<T>(promise: Promise<T>, label?: string) {
-  let result = promise;
+export default function waitForPromise(
+  target: object,
+  _key: string,
+  descriptor: PropertyDescriptor,
+  label?: string
+): PropertyDescriptor;
+export default function waitForPromise<T>(promise: Promise<T>, label?: string): Promise<T>;
+export default function waitForPromise<T>(fn: Function, label?: string): Function;
+
+export default function waitForPromise<T>(
+  ...args:
+    | [object, string, PropertyDescriptor, string?]
+    | [Promise<T>, string?]
+    | [Function, string?]
```

Comment:
Should we define a public type for these tuples?

Reply:
🤷‍♂ I'm new to typescript, so I'm not sure the pros/cons, or exactly how I would do it...

Reply:
I think we can actually leave this. It does a good job identifying the specific arguments for the 3 call signatures. Let's leave it.